### PR TITLE
Add missing trace flags to Span and Link messages

### LINKED ISSUE #382: encoding of context flags in otlp messages
[The spec for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spancontext) includes a trace flags in the context:

> TraceFlags contain details about the trace. Unlike TraceState values, TraceFlags are present in all traces. The current version of the specification only supports a single flag called [sampled](https://www.w3.org/TR/trace-context/#sampled-flag).

However, the otlp proto message has no appropriate proto field to store the flags for `message Span` and `message Link`.

I think the flags are important to be encoded in the otlp format for the following reasons:
- It is handy that when holding a link, to be able to tell if it points on a sampled or non sampled span. For example in messaging specification, a consumer may link to a producer span, but it is possible and common that the consumer is sampled and the producer is not. Currently, consumer links just store `trace_id`, `span_id` and `trace_state`. If someone tries to follow the link and cannot find the source span, there is no way to tell if it was not sampled (valid) or simply missing (lost - invalid). It is can also be expensive to follow a link (query to db) - and by knowing if a link is sampled or not one can skip this operation for processing and visualization purposes. There can also be UI value in presenting a link when it is not sampled vs sampled link.
- OTLP can be handy for storing spans, not only for transmitting them. For transmitting spans from SDKs to collector, we do not anticipate any non-sampled spans. However, someone might need to serialize and deserialize non-sampled spans (for example - to debug) in which case this data will currently be lost. I think protocol should support this kind of use-cases
- Even though flags currently only support single "sampled" flag, they can be extended in the future to hold new values. We cannot anticipate the use of these future values, and having them in the otlp model will allow using it always.

Does it make sense to add "trace_flags" field to Span and Link messages in proto? or are there any good reasons why no to include them?

### PR DESCRIPTION
**Pull request by blumamir**
fixes #382 and #166

### CODE REVIEW

**Comment by blumamir** [comment]
I now wonder if the otlp messages should also contain the [w3c version](https://www.w3.org/TR/trace-context-1/#version) so when more flags are added in the future, the receiver can differentiate between them being set to "0" vs "not set" by examining the version.
Any thoughts?
@dyladan

File: opentelemetry/proto/trace/v1/trace.proto [hunk 1]
```diff
@@ -291,6 +291,15 @@ message Span {
...
     uint32 dropped_attributes_count = 5;
+
+    // Flags, a bit field. 8 least significant bits are the trace flags as
+    // defined in W3C Trace Context specification. 24 most significant bits are reserved
+    // and must be set to 0. Readers must not assume that 24 most significant bits
+    // will be zero and must correctly mask the bits when reading 8-bit trace flag (use
+    // flags & TRACE_FLAGS_MASK).
+    // This value is required, as not setting it will be interpeted as 0 which is 
+    // a valid representation of all flags being present and set to 0.
+    fixed32 flags = 6;
```
Comment (bogdandrutu) [inline opentelemetry/proto/trace/v1/trace.proto:302]: Can you move this closer to the ids?

**Comment by bogdandrutu** [comment]
@blumamir I think we can add the version when w3c changes to v2, before that, the absence of the version means version 1.

File: opentelemetry/proto/trace/v1/trace.proto (hunk 1 above)
Comment (blumamir) [reply opentelemetry/proto/trace/v1/trace.proto:302]: No problem, but it looks like the messages are currently sorted by "field number", and moving it up will break this pattern. Are we OK with that?

**Comment by blumamir** [comment]
> @blumamir I think we can add the version when w3c changes to v2, before that, the absence of the version means version 1.

If we add it when w3c bump the version, then all existing exporters out there will not report it which will be interpreted as `0` and we are back to the original backward support problem that we are now facing.

It means that a processor will not be able to use a new flag until *all* the exporters in production are upgraded to a version that supports the new `proto` release with the w3c version field, which can be a long and painful process.

**Comment by bogdandrutu** [comment]
> It means that a processor will not be able to use a new flag until all the exporters in production are upgraded to a version that supports the new proto release with the w3c version field, which can be a long and painful process.

Isn't that true anyway? Let's assume we add the field and everyone just exports "v=1", what is the difference?

**Comment by Oberon00** [comment]
Should we maybe wait a bit with this and do it together with IsRemote once https://github.com/open-telemetry/oteps/pull/182 is merged, to minimize the chance of inconsistencies?

**Comment by tigrannajaryan** [comment]
> Should we maybe wait a bit with this and do it together with IsRemote once [open-telemetry/oteps#182](https://github.com/open-telemetry/oteps/pull/182) is merged, to minimize the chance of inconsistencies?

We can wait. If it is delayed for a long time we can also move forward with this since I think we can add new flags in backward compatible manner. For IsRemote we likely will need 2 bits to represent (Unspecified=00, IsRemote=01, IsNotRemote=01) state. Unspecified will be the default state matching the zero state of the field (which will make it backward compatible with existing OTLP versions).

**Comment by tigrannajaryan** [comment]
https://github.com/open-telemetry/oteps/pull/182 has not seen much progress. Do we want to move forward with this PR?

**Comment by jmacd** [comment]
Closing this in favor of #503. The purpose is identical, but we believe @blumamir is no longer focused on OTel so I will carry this out.

**Comment by dyladan** [comment]
@blumamir was on a break for a while but is back and active in otel again. I'm sure he probably doesn't mind work being consolidated into #503 though

**Pull request closed without merging**